Integrate the current values present the column into the column defin…

dflaven opened the pull request:
Make sure that update of the DBSchema does not break when an ENUM column contains values which are not (or no longer) part of the field definition.

dflaven committed:
Integrate the current values present the column into the column definition

bruno-ds commented:
So smart!

Is there any easy way to produce a flash message alerting the admin that the DB has such rogue values?

I think you can easily collect them using this before the array_merge:
```php
$aRogueValues = array_diff($this->GetActualValuesInDB(), $aValues);
```

Molkobain commented on core/attributedef.class.inc.php:
```diff
@@ -5206,6 +5210,28 @@ protected function GetSQLCol($bFullSpec = false)
 
+	/**
+	 * Get the list of the actual 'enum' values present in the database
+	 * @return string[]
```
Please add a `@since 3.0.0` so we know it's only available from there (if not accepted in the 3.0.0 we can always change it later before merging the PR)

dflaven replied:
Done.

Molkobain commented on core/attributedef.class.inc.php:
```diff
@@ -5188,6 +5188,10 @@ protected function GetSQLCol($bFullSpec = false)
 		}
+
+		// Preserve the values already present in the database to ease migrations
+		$aValues = array_unique(array_merge($aValues, $this->GetActualValuesInDB()));
```
I'm no expert on this part of the app. but looks like the `AttributeDefinition::GetSQLCol()` gets called in `AttributeDefinition::GetSQLColumns()` which is called in the "Datamodel viewer" page and when synchros are executed in "SynchroExecution::PrepareProcessing()" (among other less sensitive places).

Wouldn't this cause significant slowdowns on high volume DBs? Or will it use the query cache? Either way, this should be tested on big DBs, I find it a bit dangerous without an explicit cache of this somewhere at compile time.

dflaven replied:
I replaced the call in the Datamodel viewer with the static method IsBasedOnDBColumns() which seem sufficient for what it was doing.

For the data synchro, getting the values dynamically may be useful/needed anyway...

Molkobain commented:
That could indeed be a big step forward as it is a real pain in the 🍍 right now, for both the community and those using the ITSM Designer!

Just concerned about the performance consequences though.

> ...
> Is there any easy way to produce a flash message alerting the admin that the DB has such rogue values?
> ...

A flash message for the admin to see right after setup/MTP would be a good idea indeed.
It should also appear in the DB tools IMHO, as it is something to fix.

Molkobain commented:
_(Self-assigned because it has be to be assigned to someone so it can be discussed in the team meeting)_

jbostoen commented:
This would indeed be a big step forward! I've seen it several times. A proper warning would be really helpful.

dflaven committed:
Integrating review comments
- Added @since on the new method
- Replaced the call to GetSQLColumns in the schema viewer since the return values were actually not used

dflaven committed:
Trying to get the actual values... fails when the table does not exist (yet)

odain-cbd commented:
please wait before merging in develop. I reported an issue when migration from 2.7 to this current branch.

Molkobain commented:
Hello @odain-cbd , can you tell us more about the issue? Do you have a bug number?

Molkobain commented:
Discussed with @odain-cbd , the issue is fixed, we can move forward with this PR. I'll contact Denis ASAP.

Molkobain commented:
Denis is currently working on it to improve the data synchro part.

dflaven committed:
Handle Synchro Data Sources - but with the limitation of NOT taking into account the actual values inside the synchro table itself.

dflaven committed:
Fixed a missing parameter - gasp!

Molkobain commented:
Discussed with Denis, all good!
